Add useful operators and functions to FRotator

igromanru opened the pull request:
**List of changes**
- Add `UnderlayingType` to FRotator (FRotator is TRotator<double> in latest UE versions, same as FVector and FVector2D)
- Add useful operators and functions to FRotator
- In FVector and FVector2D change Scalar type to `UnderlayingType`
- Change 0.0/0.0f to simple 0. (It works with 0 with float and double without build warnings)

**Additional Info**
The code in the functions is basically copy-paste from the UE source code.   
When working with FRotators, they should be normalized in this way before they are put back into the game.  
While most standard functions that accept FRotator as a parameter perform the normalization themselves, it might be not the case for new, game specific functions.  
This is also a common trap for detection in some games when people write directly to the ControlRotation property with abnormal angles.

**Commit format**
In case you're wondering what "feat" means, my commits format is based on [Git Commits Conventional ](https://www.conventionalcommits.org/)

igromanru commented:
For better overview. This is how FRotator looks like after my changes in an SDK dump of Dead By Daylight (currently UE v5.2)
```cpp
// ScriptStruct CoreUObject.Rotator
// 0x0018 (0x0018 - 0x0000)
struct FRotator final
{
public:
	using UnderlayingType = double;

	double                                        Pitch;                                             // 0x0000(0x0008)(Edit, BlueprintVisible, ZeroConstructor, SaveGame, IsPlainOldData, NoDestructor, HasGetValueTypeHash, NativeAccessSpecifierPublic)
	double                                        Yaw;                                               // 0x0008(0x0008)(Edit, BlueprintVisible, ZeroConstructor, SaveGame, IsPlainOldData, NoDestructor, HasGetValueTypeHash, NativeAccessSpecifierPublic)
	double                                        Roll;                                              // 0x0010(0x0008)(Edit, BlueprintVisible, ZeroConstructor, SaveGame, IsPlainOldData, NoDestructor, HasGetValueTypeHash, NativeAccessSpecifierPublic)

public:
	static UnderlayingType ClampAxis(UnderlayingType Angle)
	{
		Angle = std::fmod(Angle, (UnderlayingType)360.0);
		if (Angle < (UnderlayingType)0)
			Angle += (UnderlayingType)360.0;
		return Angle;
	}
	static UnderlayingType NormalizeAxis(UnderlayingType Angle)
	{
		Angle = ClampAxis(Angle);
		if (Angle > (UnderlayingType)180.0)
			Angle -= (UnderlayingType)360.0;
		return Angle;
	}

	FRotator& Clamp()
	{
		Pitch = ClampAxis(Pitch);
		Yaw = ClampAxis(Yaw);
		Roll = ClampAxis(Roll);
		return *this;
	}
	FRotator& Normalize()
	{
		Pitch = NormalizeAxis(Pitch);
		Yaw = NormalizeAxis(Yaw);
		Roll = NormalizeAxis(Roll);
		return *this;
	}
	FRotator& operator*=(UnderlayingType Scalar)
	{
		*this = *this * Scalar;
		return *this;
	}
	FRotator& operator*=(const FRotator& Other)
	{
		*this = *this * Other;
		return *this;
	}
	FRotator& operator+=(const FRotator& Other)
	{
		*this = *this + Other;
		return *this;
	}
	FRotator& operator-=(const FRotator& Other)
	{
		*this = *this - Other;
		return *this;
	}
	FRotator& operator/=(UnderlayingType Scalar)
	{
		*this = *this / Scalar;
		return *this;
	}
	FRotator& operator/=(const FRotator& Other)
	{
		*this = *this / Other;
		return *this;
	}

	FRotator GetNormalized() const
	{
		FRotator rotator = *this;
		rotator.Normalize();
		return rotator;
	}
	bool IsZero() const
	{
		return ClampAxis(Pitch) == 0 && ClampAxis(Yaw) == 0 && ClampAxis(Roll) == 0;
	}
	bool operator!=(const FRotator& Other) const
	{
		return Pitch != Other.Pitch || Yaw != Other.Yaw || Roll != Other.Roll;
	}
	FRotator operator*(UnderlayingType Scalar) const
	{
		return { Pitch * Scalar, Yaw * Scalar, Roll * Scalar };
	}
	FRotator operator*(const FRotator& Other) const
	{
		return { Pitch * Other.Pitch, Yaw * Other.Yaw, Roll * Other.Roll };
	}
	FRotator operator+(const FRotator& Other) const
	{
		return { Pitch + Other.Pitch, Yaw + Other.Yaw, Roll + Other.Roll };
	}
	FRotator operator-(const FRotator& Other) const
	{
		return { Pitch - Other.Pitch, Yaw - Other.Yaw, Roll - Other.Roll };
	}
	FRotator operator/(UnderlayingType Scalar) const
	{
		if (Scalar == 0)
			return *this;
	
		return { Pitch / Scalar, Yaw / Scalar, Roll / Scalar };
	}
	FRotator operator/(const FRotator& Other) const
	{
		if (Other.Pitch == 0 || Other.Yaw == 0 || Other.Roll == 0)
			return *this;
	
		return { Pitch / Other.Pitch, Yaw / Other.Yaw, Roll / Other.Roll };
	}
	bool operator==(const FRotator& Other) const
	{
		return Pitch == Other.Pitch && Yaw == Other.Yaw && Roll == Other.Roll;
	}
};
```

Fischsalat commented:
Thank you for the pullrequest! Time is currently scarce for me, so I will likely review and test the PR in a week or so.

igromanru commented:
No pressure or anything, but it would be nice to get feedback.
I mean there isn't much you have to do, just clone my fork, compile it and inject in any game that doesn't require `MoreGameSupport` branch, like current Dead by Daylight (UE 5.2.1).  
The code changes are pretty straightforward.

igromanru commented:
Alright, but I changed it only in the code I added.  
I can refactor code that was there before, for FVector and others, if you want, but I would rather do it in another RP to keep it clean.

igromanru commented:
I've changed it already. Can I just keep it this way now and we can refactor FVector and FVector2D later?

Fischsalat commented:
Yeah , sure.

Fischsalat commented:
@igromanru could you also add a constructor/assignment operator to FRotator/FVector2D? Thanks.

igromanru commented:
Alright
